torchgen/model.py

```diff
@@ -838,9 +882,93 @@ def from_dict(
+            # The default grouping logic in signature() actually already does this,
+            # so we can piggy-back off it (but we still want return names)
+            func = f.func.signature(keep_return_names=True).with_name(
+                f.func.name.remove_inplace().with_overload(
+                    "functional"
+                    if not f.func.name.overload_name
+                    else f"{f.func.name.overload_name}_functional"
+                )
+            )
+        elif k == SchemaKind.out:
+            # We generate out= ops mostly just so that we can pair up NativeFunctions into groups easily,
+            # but at least today, there is no good reason to actually use them.
+            # we'll generate a dispatcher entry for them, but won't actually register any kernels for them.
+            gets_composite_kernel = False
+            if f.func.kind() == SchemaKind.inplace:
+                func = f.func.self_to_out_signature()
+            elif f.func.kind() == SchemaKind.mutable:
+                func = f.func.mutable_to_out_signature()
+            else:
+                raise AssertionError(
+                    "We only bother generating out= functions from either inplace or mutable variants"
+                )
+        else:
+            raise AssertionError(
+                "We currently only generate either functional or out= NativeFunctions"
+            )
+
+        if gets_composite_kernel:
+            backend_metadata = {
+                DispatchKey.CompositeExplicitAutograd: {
+                    func.name: BackendMetadata(cpp.name(func), structured=False)
+                }
+            }
+        else:
+            backend_metadata = {}
+
+        return (
+            NativeFunction(
+                func=func,
+                use_const_ref_for_mutable_tensors=f.use_const_ref_for_mutable_tensors,
+                # These generated fn's aren't meant to be user friendly- don't generate methods.
+                variants=set([Variant.function]),
+                structured=False,
+                structured_delegate=None,
+                structured_inherits=None,
+                precomputed=None,
+                ufunc_inner_loop={},
+                manual_kernel_registration=False,
+                manual_cpp_binding=False,
+                python_module=None,
+                category_override=None,
+                device_guard=False,
+                device_check=DeviceCheckType.NoCheck,
+                loc=f.loc,
+                cpp_no_default_args=set(),
+                is_abstract=f.is_abstract,
+                has_composite_implicit_autograd_kernel=False,
+                has_composite_explicit_autograd_kernel=gets_composite_kernel,
+                # Every generated NativeFunction gets a "generated" tag, so it's easy to tell
+                # which NativeFunction objects did not come directly from native_functions.yaml.
+                tags=set(["generated"]),
```

Comment:
🤔

Reply:
Tags shouldn't be inherited right now, however we should once we have two categories of tags as described below.

It might make sense to have two category of tags: ones that are generally applicable to all overloads for an op (e.g., unary pointwise) and others that are specifically applicable to some overloads (for e.g., inplace_view). The former category can be thought of as a set of tags that are associated with the mathematical semantics of the actual operation in question, and latter can be thought of as a tag indicative of its programatic implementation (mutability for instance). 

This could be implemented by just adding another entry for native_functions (packet_tags or something). However one thing to keep in mind is that we will need to add additional API to `.def()` to indicate these tags if we go down this route. Perhaps, adding this information as a metadata in tags.yaml could be a better route in that regard.

Reply:
No inheritance is a problem for autogen because there is literally no other way to specify the tags for an autogenerated operator. I think we should distinguish between "different overloads" (which can be quite different) and "operators in the same native functions group" (which really should be semantically equivalent).